common/amount.h

```diff
@@ -0,0 +1,142 @@
+#ifndef LIGHTNING_COMMON_AMOUNT_H
+#define LIGHTNING_COMMON_AMOUNT_H
+#include "config.h"
+#include <ccan/build_assert/build_assert.h>
+#include <ccan/short_types/short_types.h>
+#include <ccan/tal/tal.h>
+#include <stdbool.h>
+
+#define MSAT_PER_SAT ((u64)1000)
+#define SAT_PER_BTC ((u64)100000000)
+#define MSAT_PER_BTC (MSAT_PER_SAT * SAT_PER_BTC)
+
+/* Use these to wrap amounts, for typesafety.  Please use ops where possible,
+ * rather than accessing the members directly. */
+struct amount_sat {
+	/* Amount in millisatoshis. */
+	u64 satoshis;
+};
+
+struct amount_msat {
+	/* Amount in millisatoshis. */
+	u64 millisatoshis;
+};
+
+/* For constants only: others must be built from primitives! */
+#if HAVE_BUILTIN_CONSTANT_P
+#define AMOUNT_MUST_BE_CONST(c) BUILD_ASSERT_OR_ZERO(IS_COMPILE_CONSTANT(c))
+#else
+#define AMOUNT_MUST_BE_CONST(c) 0
+#endif
+
+#define AMOUNT_MSAT(constant)						\
+	((struct amount_msat){(constant) + AMOUNT_MUST_BE_CONST(constant)})
+
+#define AMOUNT_SAT(constant)						\
+	((struct amount_sat){(constant) + AMOUNT_MUST_BE_CONST(constant)})
+
+/* You may not always be able to convert satoshis->millisatoshis. */
+WARN_UNUSED_RESULT bool amount_sat_to_msat(struct amount_msat *msat,
+					   struct amount_sat sat);
```

Comment:
It's usual to pass structs by pointer rather than copy, yes.  But it's (really) just a u64, so copying is basically free, and passing by copy is easier to chain functions together which return a struct.  esp AMOUNT_SAT(1000) works.